test/core/event_engine/posix/posix_endpoint_test.cc

```diff
@@ -0,0 +1,352 @@
+// Copyright 2022 gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/core/lib/event_engine/posix_engine/posix_endpoint.h"
+
+#include <fcntl.h>
+#include <poll.h>
+
+#include <chrono>
+#include <memory>
+#include <random>
+#include <string>
+#include <thread>
+
+#include <gtest/gtest.h>
+
+#include "absl/strings/str_cat.h"
+
+#include <grpc/event_engine/event_engine.h>
+
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/event_engine/channel_args_endpoint_config.h"
+#include "src/core/lib/event_engine/posix_engine/event_poller.h"
+#include "src/core/lib/event_engine/posix_engine/event_poller_posix_default.h"
+#include "src/core/lib/event_engine/posix_engine/posix_engine.h"
+#include "src/core/lib/event_engine/posix_engine/posix_engine_closure.h"
+#include "src/core/lib/event_engine/posix_engine/tcp_socket_utils.h"
+#include "src/core/lib/gprpp/dual_ref_counted.h"
+#include "src/core/lib/gprpp/global_config.h"
+#include "src/core/lib/gprpp/notification.h"
+#include "src/core/lib/iomgr/port.h"
+#include "test/core/event_engine/posix/posix_engine_test_utils.h"
+#include "test/core/event_engine/test_suite/event_engine_test_utils.h"
+#include "test/core/event_engine/test_suite/oracle_event_engine_posix.h"
+#include "test/core/util/port.h"
+
+GPR_GLOBAL_CONFIG_DECLARE_STRING(grpc_poll_strategy);
+
+namespace grpc_event_engine {
+namespace posix_engine {
+
+namespace {
+
+using ::grpc_event_engine::experimental::ChannelArgsEndpointConfig;
+using ::grpc_event_engine::experimental::EventEngine;
+using ::grpc_event_engine::experimental::GetNextSendMessage;
+using ::grpc_event_engine::experimental::Poller;
+using ::grpc_event_engine::experimental::PosixEventEngine;
+using ::grpc_event_engine::experimental::PosixOracleEventEngine;
+using ::grpc_event_engine::experimental::URIToResolvedAddress;
+using ::grpc_event_engine::experimental::WaitForPendingTasks;
+using Endpoint = ::grpc_event_engine::experimental::EventEngine::Endpoint;
+using Listener = ::grpc_event_engine::experimental::EventEngine::Listener;
+using namespace std::chrono_literals;
+
+constexpr int kMinMessageSize = 1024;
+constexpr int kNumConnections = 10;
+constexpr int kNumExchangedMessages = 100;
+std::atomic<int> g_num_active_connections{0};
+
+EventEngine* GetOracleEE() {
+  static EventEngine* oracle_ee = new PosixOracleEventEngine();
```

Comment:
Shouldn't TSAN complain about threads leaked off the end of the process if this engine is never destroyed?